### DISCUSSION TITLE
[TlsInterception] How to parse JSON (chunked, brotli compressed) and modify it ?

### DISCUSSION THREAD

**Discussion by ghostlexly** [abhinavsingh/proxy.py#955] [Q&A] [answered]
Hello, i'm trying a  man-in-the-middle attack, and i can't find a way to change the JSON response.

here is my code

````
    def handle_upstream_chunk(self, chunk: memoryview) -> memoryview:
        self.response.parse(chunk.tobytes())

        if self.response.body_expected:
            self.response.body = b'\nTEST'
            self.client.queue(memoryview(self.response.build_response()))
            return okResponse(content=b'Hello from man in the middle')

        return chunk
````

I can't read the body as JSON and can't find a way to replace or search for a text inside..

**Comment by abhinavsingh** [comment, marked as answer]
As `handle_upstream_chunk` suggests, incoming data is a `chunk`.  You must assemble all the chunks before you can make sense of them together.  Incoming chunks can be assembled and parsed using `ChunkParser` class.  See https://github.com/abhinavsingh/proxy.py/blob/develop/proxy/http/parser/chunk.py#L27-L35

Once you have all chunks assembled, you should be able to parse that data as json (if it is a valid json).  You can ensure it's json by checking `Content-Type` for responses

**Reply by ghostlexly** [reply]
It's working but all the data is encrypted, see:

````
wF[xPi=2}qTC;ſ#DQob̴exsek쌲*1PF]P>y",@jE7mȎ0Q!9|[\`(c** ?
]^4P&g]0$|։ZJnAG!.ݎax+wM,QBM~y}яLⷒ-+vסS_,IOʟ)QcJ~JB3e7]Eҝri[pdX"^䞋1V4gFS#?[%?mZ>V<姣߳?7d�||ӸP/Ɗ_-$[@_3·D8?wI9޽]FSgv}iDN[q$YK痊QnLi僼y^\p+*LZ"WeXoOy7$ABs1!5fwQ1`˥F7J_nx'
iK@?l
jі32voM%
}æglejpTNCc;)B3*T0,ڌOխr`E)'jQuD;1tI[*mbÃ"FV]Lj@~3&B2'yRf=r߯Y"YE5ȎbIql4{ֵEdlQM0+DBG+!..mf\(^(zYzZ;qma!J܆r៩=dFR1@RAzC
m}vbT-OTMI#f@   Ҡ;g!k,qϹaJ
eRʭ7Sy\˖$ EdW]R;/v4ej@fv2sp]0G
                              L1.+㇥NRl'#2
k&-!cgS i3q
            CIo!bä5|AT
",-Q)+P>\\            YmJ_7/
          u:KXSsE֊zhHE:ۄR~oM;}ڎVN%yk"Jj+e?bHm}?*u_ C}{_~0/?E^
[U-ɋ>%HGk*E2GZD&$N2
````

I'm running the proxy with this command :

````
proxy --hostname 0.0.0.0 --plugins proxy.plugin.FilterByURLRegexPlugin  --ca-key-file /root/proxy.py-develop/ca-key.pem --ca-cert-file /root/proxy.py-develop/ca-cert.pem --ca-signing-key-file /root/proxy.py-develop/ca-signing-key.pem
````

I'm giving him the certificates but the chunks are still encrypted

**Reply by abhinavsingh** [reply]
Well who knows this is just compressed data.  Check response headers to see what compression algorithm was applied.  Then decompress it.

At some point we must add a utility within `proxy.py` core for users to achieve all of this (for chunk responses) within a single function call.  Let me know if it works, then we can think how to generically add it.

**Reply by ghostlexly** [reply]
Here is the headers:

````
{b'vary': (b'Vary', b'Accept-Encoding, Origin'), b'content-encoding': (b'Content-Encoding', b'br'), b'accept-ranges': (b'Accept-Ranges', b'bytes'), b'content-type': (b'Content-Type', b'text/javascript'), b'cross-origin-resource-policy': (b'Cross-Origin-Resource-Policy', b'cross-origin'), b'cross-origin-opener-policy-report-only': (b'Cross-Origin-Opener-Policy-Report-Only', b'same-origin; report-to="youtube"'), b'report-to': (b'Report-To', b'{"group":"youtube","max_age":2592000,"endpoints":[{"url":"https://csp.withgoogle.com/csp/report-to/youtube"}]}'), b'content-length': (b'Content-Length', b'1294456'), b'date': (b'Date', b'Fri, 07 Jan 2022 11:31:10 GMT'), b'expires': (b'Expires', b'Sat, 07 Jan 2023 11:31:10 GMT'), b'last-modified': (b'Last-Modified', b'Fri, 07 Jan 2022 08:14:31 GMT'), b'x-content-type-options': (b'X-Content-Type-Options', b'nosniff'), b'server': (b'Server', b'sffe'), b'x-xss-protection': (b'X-XSS-Protection', b'0'), b'cache-control': (b'Cache-Control', b'public, max-age=31536000'), b'age': (b'Age', b'280735'), b'alt-svc': (b'Alt-Svc', b'h3=":443"; ma=2592000,h3-29=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"')}
````

Can't find any encoding header :/
I'm trying to change the netflix.com's homepage text for MITM attack test purposes

Do you have any code example of https mitm ? i have found one in the plugins folder but it's doesn't read the page's content

Thank you

**Reply by abhinavsingh** [reply]
`b'content-encoding': (b'Content-Encoding', b'br')`

`br` stands for https://en.wikipedia.org/wiki/Brotli

**Reply by abhinavsingh** [reply]
Simply find a library which supports `brotli` compression/decompression and use it :D.   Once decompressed, as indicated in headers, content type is `b'content-type': (b'Content-Type', b'text/javascript')` and not `JSON`, as you are trying to parse.

**Reply by abhinavsingh** [reply]
> I'm trying to change the netflix.com's homepage text for MITM attack test purposes
> 
> Do you have any code example of https mitm ? i have found one in the plugins folder but it's doesn't read the page's content

Sorry, I use `TlsInterception` for analysis, but never for `MITM`.  I added support for `TlsInterception` mainly for analysis purposes.  That's another thing that it also allows `MITM`.  Support was added because `proxy.py` users requested for it.

**Reply by ghostlexly** [reply]
Thanks for the br, i was looking for gzip lol

here is my final code :

````
    import brotli

    def handle_upstream_chunk(self, chunk: memoryview) -> memoryview:
        self.response.parse(chunk.tobytes())

        if type(self.response.body) is bytes:
            if self.response.chunk_parser.state == chunkParserStates.COMPLETE:
                print(self.response.headers)
                print(brotli.decompress(self.response.body))

        return chunk
````

But now i have `"Decompression error: b'CL_SPACE'"` and `"Decompression error: b'HUFFMAN_SPACE'"` errors on all `brotli` requests.  I think it's because the chunks are incomplete. I don't understand why.

**Reply by abhinavsingh** [reply]
Sorry I am unsure what's happening here either.  But if chunk parser has reached `COMPLETE` state, we can be rest assured that we have received expected data.  How does `len(self.response.body)` compares to `Content-Length` header value?

**Reply by ghostlexly** [reply]
Where can i find the content-length value ?
i don't have it in the headers:

````
    def handle_upstream_chunk(self, chunk: memoryview) -> memoryview:
        self.response.parse(chunk.tobytes())

        if type(self.response.body) is bytes:
            if self.response.chunk_parser.state == chunkParserStates.COMPLETE:
                print(self.response.headers)
                print(len(self.response.body))

        return chunk
````

````
{b'content-type': (b'Content-Type', b'application/javascript; charset=utf-8'), b'transfer-encoding': (b'Transfer-Encoding', b'chunked'), b'connection': (b'Connection', b'keep-alive'), b'date': (b'Date', b'Mon, 10 Jan 2022 19:07:31 GMT'), b'server': (b'Server', b'nginx'), b'etag': (b'ETag', b'W/"1ac59-zzvgiaPS/0qnPUAtRCdiNA"'), b'cache-control': (b'Cache-Control', b'public, max-age=120'), b'strict-transport-security': (b'Strict-Transport-Security', b'max-age=31536000; includeSubDomains'), b'content-encoding': (b'Content-Encoding', b'gzip'), b'vary': (b'Vary', b'Accept-Encoding'), b'x-cache': (b'X-Cache', b'Hit from cloudfront'), b'via': (b'Via', b'1.1 9d2c93ece5a5ccb2b5952a40f7502a04.cloudfront.net (CloudFront)'), b'x-amz-cf-pop': (b'X-Amz-Cf-Pop', b'CDG3-C1'), b'x-amz-cf-id': (b'X-Amz-Cf-Id', b'lEZOFQUz2aTn2bRuqnSw4PLdWWphmKazRB_1poHma7LQsWuWe2tlDw=='), b'age': (b'Age', b'46')}

body length is : 42849
````

I attach my full plugin code here if you wanna do some tests on your side.
Big thank you, it's an important project for my company

[filter_by_url_regex.txt](https://github.com/abhinavsingh/proxy.py/files/7841328/filter_by_url_regex.txt)

**Reply by abhinavsingh** [reply]
You had a `content-length` header in previous example.  Your current example don't. 

<img width="830" alt="Screen Shot 2022-01-11 at 1 24 52 AM" src="https://user-images.githubusercontent.com/126065/148830364-2f956033-393d-458d-ae2f-04caad325200.png">

Irrespective this is an issue unrelated to `proxy.py`.  You will run into variety of variations in the wild.  `proxy.py` only provides you an ability to get hold of the packets.  Further you must use your favorite library to parse/split/modify them.

I tried following, so as such I see no issue with receiving brotli compressed data (via `proxy.py`)

```console
curl -x localhost:50660 http://httpbin.org/brotli | brotli -d -                                            ─╯
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   173  100   173    0     0    208      0 --:--:-- --:--:-- --:--:--   210
{
  "brotli": true, 
  "headers": {
    "Accept": "*/*", 
    "Host": "httpbin.org", 
    "User-Agent": "curl/7.77.0", 
    "X-Amzn-Trace-Id": "Root=1-61dc8df1-14c6329d103c4239605114a8"
  }, 
  "method": "GET", 
}
```
